oss-internship-2020/curl/tests/test_utils.cc

```diff
@@ -0,0 +1,287 @@
+  } else if (curl_code.value() != CURLE_OK) {
+    return absl::UnavailableError(
+        "curl_easy_setopt_ptr returned with the error code " +
+        curl_code.value());
+  }
+
+  // Set port
+  curl_code = api_->curl_easy_setopt_long(curl_.get(), CURLOPT_PORT, port_);
+  if (!curl_code.ok()) {
+    return curl_code.status();
+  } else if (curl_code.value() != CURLE_OK) {
+    return absl::UnavailableError(
+        "curl_easy_setopt_long returned with the error code " +
+        curl_code.value());
+  }
+
+  // Generate pointer to the WriteToMemory callback
+  void* function_ptr;
+  absl::Status symbol =
+      sandbox_->rpc_channel()->Symbol("WriteToMemoryTests", &function_ptr);
+  if (!symbol.ok()) {
+    return symbol;
+  }
+  sapi::v::RemotePtr remote_function_ptr(function_ptr);
+
+  // Set WriteToMemory as the write function
+  curl_code = api_->curl_easy_setopt_ptr(curl_.get(), CURLOPT_WRITEFUNCTION,
+                                         &remote_function_ptr);
+  if (!curl_code.ok()) {
+    return curl_code.status();
+  } else if (curl_code.value() != CURLE_OK) {
+    return absl::UnavailableError(
+        "curl_easy_setopt_ptr returned with the error code " +
+        curl_code.value());
+  }
+
+  // Pass memory chunk object to the callback
+  curl_code = api_->curl_easy_setopt_ptr(curl_.get(), CURLOPT_WRITEDATA,
+                                         chunk_.PtrBoth());
+  if (!curl_code.ok()) {
+    return curl_code.status();
+  } else if (curl_code.value() != CURLE_OK) {
+    return absl::UnavailableError(
+        "curl_easy_setopt_ptr returned with the error code " +
+        curl_code.value());
+  }
+
+  return absl::OkStatus();
+}
+
+absl::Status CurlTestUtils::CurlTestTearDown() {
+  // Cleanup curl
+  return api_->curl_easy_cleanup(curl_.get());
+}
+
+absl::StatusOr<std::string> CurlTestUtils::PerformRequest() {
+  // Perform the request
+  absl::StatusOr<int> curl_code = api_->curl_easy_perform(curl_.get());
+  if (!curl_code.ok()) {
+    return curl_code.status();
+  } else if (curl_code.value() != CURLE_OK) {
+    return absl::UnavailableError(
+        "curl_easy_perform returned with the error code " + curl_code.value());
+  }
+
+  // Get pointer to the memory chunk
+  sapi::v::GenericPtr remote_ptr;
+  remote_ptr.SetRemote(
+      &(static_cast<sapi::LenValStruct*>(chunk_.GetRemote()))->data);
+  absl::Status transfer = sandbox_->TransferFromSandboxee(&remote_ptr);
+  if (!transfer.ok()) {
+    return transfer;
+  }
+  void* chunk_ptr = (void*)remote_ptr.GetValue();
+
+  // Get the string
+  absl::StatusOr<std::string> response =
+      sandbox_->GetCString(sapi::v::RemotePtr(chunk_ptr));
+  if (!response.ok()) {
+    return response.status();
+  }
+
+  return response.value();
+}
+
+// Read the socket until str is completely read
+std::string ReadUntil(const int socket, const std::string& str,
+                      const size_t max_request_size) {
+  char buf[max_request_size] = {};
```

Comment:
Variable length arrays are not standard.
https://google.github.io/styleguide/cppguide.html#Nonstandard_Extensions

Read directly into std::string.
First reserve max_request_size then just push_back the read char.
Also checking for suffix can be simplified with Abseil (https://github.com/abseil/abseil-cpp) absl::EndsWith - this is a core Google C++ library and would also simplify other stuff.